pkg/manager/member/common_store_failover.go

```diff
@@ -33,29 +35,69 @@ type StoreAccess interface {
 	GetMemberType() v1alpha1.MemberType
 	GetMaxFailoverCount(tc *v1alpha1.TidbCluster) *int32
 	GetStores(tc *v1alpha1.TidbCluster) map[string]v1alpha1.TiKVStore
+	GetStore(tc *v1alpha1.TidbCluster, storeID string) (v1alpha1.TiKVStore, bool)
 	SetFailoverUIDIfAbsent(tc *v1alpha1.TidbCluster)
 	CreateFailureStoresIfAbsent(tc *v1alpha1.TidbCluster)
 	GetFailureStores(tc *v1alpha1.TidbCluster) map[string]v1alpha1.TiKVFailureStore
+	GetFailureStore(tc *v1alpha1.TidbCluster, storeID string) (v1alpha1.TiKVFailureStore, bool)
 	SetFailureStore(tc *v1alpha1.TidbCluster, storeID string, failureStore v1alpha1.TiKVFailureStore)
 	ClearFailStatus(tc *v1alpha1.TidbCluster)
 	GetStsDesiredOrdinals(tc *v1alpha1.TidbCluster, excludeFailover bool) sets.Int32
+	IsHostDownForFailurePod(tc *v1alpha1.TidbCluster) bool
 }
 
-// commonStoreFailover handles the common failover logic for TiKV and TiFlash
+// commonStoreFailover has the common logic to handle the failover of TiKV and TiFlash store
 type commonStoreFailover struct {
-	deps        *controller.Dependencies
-	storeAccess StoreAccess
+	deps            *controller.Dependencies
+	storeAccess     StoreAccess
+	failureRecovery commonStatefulFailureRecovery
 }
 
 var _ Failover = (*commonStoreFailover)(nil)
 
 func (sf *commonStoreFailover) Failover(tc *v1alpha1.TidbCluster) error {
+	// If store downtime exceeds failover time period then create a FailureStore
+	// The failure recovery of down store follows this timeline:
+	// If HostDown is not set then detect node failure for the pod and set HostDown
+	// If HostDown is set then force restart pod once
+	// If HostDown is set and Store is Down then delete Store after some gap from the time of pod restart
+	// If HostDown is set and Store has been removed or become Tombstone then remove PVC and set StoreDeleted
+
+	if err := sf.failureRecovery.CheckHostDownAndRestartPod(tc); err != nil {
+		if controller.IsIgnoreError(err) {
+			return nil
+		}
+		return err
+	}
+
 	if err := sf.tryMarkAStoreAsFailure(tc); err != nil {
 		if controller.IsIgnoreError(err) {
 			return nil
 		}
 		return err
 	}
+
+	if canAutoFailureRecovery(tc) {
+		// If the store has not come back Up in some time after pod was restarted, then delete store and remove failure PVC
+		failureStores := sf.storeAccess.GetFailureStores(tc)
+		for storeId := range failureStores {
+			failureStore := failureStores[storeId]
```

Comment:
In general I think iterating over k,v in map is not encouraged (esp. in case if the map value is a pointer). But it is safe to do that here since failureStores is a map with an object and not pointer.

Reply:
Fixed.